File: scraping/platforms/line_today.py
# THIS WORKS :D
import re
from urllib.request import urlopen, Request
import chardet
from bs4 import BeautifulSoup
import json
import psycopg2
import pandas as pd
import dotenv
import os
import gzip
import io
import argparse

parser = argparse.ArgumentParser(description="A LINE Today Scraper.")
parser.add_argument("-s", "--slug", type=str, help="The article URL like: oqmazXP")
args = parser.parse_args()

if not args.slug:
    print("No Slug entered, please use -s oqmazXP as a demo.")
    exit(1)

# Load environment variables from .env file
dotenv.load_dotenv()

headers = {
    #'User-Agent': 'NewsSceraperBot/1.0 (https://github.com/hpware/news-analyze) (A little note: If you see this, It means that your website is being scraped by other people, not the user hpware. Please keep that in mind at don't spam issues, I can't fix it.)',
    'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
    'Accept': '*',
    'Accept-Language': 'zh-TW,zh;q=0.9,en-US;q=0.8,en;q=0.7',
    'Accept-Encoding': 'gzip, deflate, br',
    'Connection': 'keep-alive',
    'Sec-Fetch-Dest': 'document',
    'Sec-Fetch-Mode': 'navigate',
    'Sec-Fetch-Site': 'same-origin',
    'Cache-Control': 'max-age=0',
}


try:
    url = "https://today.line.me/tw/v2/article/" + args.slug
    req = Request(url, headers=headers)
    response = urlopen(req)
    if response.info().get('Content-Encoding') == 'gzip':
        print("GZIP")
        gzip_file = gzip.GzipFile(fileobj=io.BytesIO(response.read()))
        html = gzip_file.read().decode('utf-8')
    else:
        print("Not GZIP")
        html = response.read().decode('utf-8')


    soup = BeautifulSoup(html, "html.parser")

    # Extract content
    title = soup.find('h1', class_="entityTitle")
    title_text = title.text.strip() if title else "No title found"

    article = soup.find('article', class_="news-content")
    paragraph = article.text.strip() if article else ""

    # Print results
    print(f"Title: {title_text}")
    print(f"Content: {paragraph}")

except Exception as e:
    print(f"Error: {str(e)}")

File: server/scrape/line_today.ts
import * as cheerio from "cheerio";

function findTime(timeText: string) {
  const now = new Date();

  const hourMatch = timeText.match(/(\d+)小時前/);
  const dayMatch = timeText.match(/(\d+)天前/);
  const minuteMatch = timeText.match(/(\d+)分鐘前/);
  if (hourMatch) {
    const hoursAgo = parseInt(hourMatch[1]);
    return new Date(now.getTime() - hoursAgo * 60 * 60 * 1000);
  } else if (dayMatch) {
    const daysAgo = parseInt(dayMatch[1]);
    return new Date(now.getTime() - daysAgo * 24 * 60 * 60 * 1000);
  } else if (minuteMatch) {
    const minutesAgo = parseInt(minuteMatch[1]);
    return new Date(now.getTime() - minutesAgo * 60 * 1000);
  }
  
  return null;
}

async function lineToday(slug: string) {
  const url = "https://today.line.me/tw/v2/article/" + slug;
  const fetchPageCode = await fetch(url, {
    headers: {
      "User-Agent":
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
      Accept: "*",
      "Accept-Language": "zh-TW,zh;q=0.9,en-US;q=0.8,en;q=0.7",
      "Accept-Encoding": "gzip, deflate, br",
      Connection: "keep-alive",
      "Sec-Fetch-Dest": "document",
      "Sec-Fetch-Mode": "navigate",
      "Sec-Fetch-Site": "same-origin",
      "Cache-Control": "max-age=0",
    },
  });
  // 幹 又忘了 await
  const data = await fetchPageCode.text();
  // 加 await? no.
  // AHHH I NEED TO CHANGE TO SOMETHING ELSE.
  const html = cheerio.load(data);
  const title = html("h1.entityTitle")
    .text()
    .replaceAll("\n", "")
    .replace("  ", "");
  const paragraph = <any[]>[];
  const images = <any[]>[];
  html("article.news-content")
    .contents()
    .each((i, element) => {
      if (element.type === "tag" && element.tagName === "figure") {
        const imgSrc = html(element).find("img").attr("src");
        if (imgSrc) {
          images.push(imgSrc);
        }
      } else if (element.type === "tag" && element.tagName === "p") {
        const text = html(element).text().trim();
        if (text) {
          paragraph.push(text);
        }
      }
    });
  const newsOrgdir = html("h4.entityPublishInfo-publisher")
    .text()
    .replaceAll("\n", "")
    .replaceAll(" ", "");

  let author = "";
  const authorInfo = html("span.entityPublishInfo-meta-info")
    .text()
    .replace(/更新.*發布.*•/g, "")
    .replaceAll("\n", "")
    .replaceAll(" ", "");
  if (/更新.*發布.*/.test(authorInfo)) {
    author = "未知";
  } else {
    author = authorInfo;
  }
  const orgAuthorDateData = html("span.entityPublishInfo-meta-info").text()
  const updateMatch = orgAuthorDateData.match(/更新於\s*([^•]+)/);
  const publishMatch = orgAuthorDateData.match(/發布於\s*(.+)$/);
let updatedAt: Date | null = null;
  if (updateMatch) {
    updatedAt = findTime(updateMatch[1].trim());
  }
let publishedAt: Date | null = null;
  if (publishMatch) {
    publishedAt = findTime(publishMatch[1].trim());
  }
  return {
    title: title,
    paragraph: paragraph,
    origin: newsOrgdir,
    author: author,
    images: images,
    updateat: updatedAt,
    publishedat: publishedAt
  };
}

export default lineToday;
